#!/bin/bash
# bam_to_fastq2 0.0.1
# Generated by dx-app-wizard.

main() {

    echo "Value of bam: '$bam'"
    echo "Value of params: '$params'"

    # boilerplate stuff needed to get dnanexus samtools running in the ubuntu
    # 12.04 execution environment
    rm /etc/apt/apt.conf.d/99dnanexus
    add-apt-repository -y ppa:ubuntu-toolchain-r/test
    apt-get update
    apt-get install -y gcc-4.8 g++-4.8
    update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-4.8 50
    update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-4.8 50
    bzip2 -d /samtools.bz2
    chmod +x /samtools

    # mount dxfs
    mkdir -p /cloud/workspace
    dx-mount /cloud/workspace

    # show bam headers
    echo "+------------+"
    echo "| BAM HEADER |"
    echo "+------------+"
    bam_fn=$(dx describe --name $(dx-jobutil-parse-link "$bam"))
    bam_fn="/cloud/workspace/$bam_fn"
    /samtools view -H "$bam_fn"

    echo "+--------------+"
    echo "| RECORD COUNT |"
    echo "+--------------+"
    dx cat $(dx-jobutil-parse-link "$bam") | tee input.bam | /samtools view -c -
    
    echo "+----------------------------+"
    echo "| RECORD COUNT BY READ GROUP |"
    echo "+----------------------------+"
    /samtools view input.bam | grep --only-matching "RG:Z:[0-9]" | sort | uniq -c
    
    echo "+-----------------------------+"
    echo "| ALIGNED COUNT BY READ GROUP |"
    echo "+-----------------------------+"
    /samtools view -F 4 input.bam | grep --only-matching "RG:Z:[0-9]" | sort | uniq -c
    
    echo "+-------------------------------+"
    echo "| UNALIGNED COUNT BY READ GROUP |"
    echo "+-------------------------------+"
    /samtools view -f 4 input.bam | grep --only-matching "RG:Z:[0-9]" | sort | uniq -c
}
