Whole-genome VCF Quality Control
========================================================

Load variants on the canonical chromosomes (b37/hs37d5 chromosome naming assumed):

```{r message=FALSE}
library(VariantAnnotation)
library(ggplot2)
vcf <- readVcf("/input.vcf.gz","hs37d5")
chroms <- c(as.character(1:22),"X","Y")
vcf <- keepSeqlevels(vcf,chroms)
```

## Quality distribution

```{r fig.width=6, fig.height=4, warning=FALSE}
qplot(qual(vcf), stat="ecdf", geom="step", xlim=c(0,2000), xlab="QUAL", ylab="CDF")
```

## Coverage distribution

```{r fig.width=6, fig.height=4, warning=FALSE}
dp.dither <- runif(nrow(vcf),-0.5,0.5)
qplot(info(vcf)$DP+dp.dither, stat="ecdf", geom="step", xlim=c(0,64), xlab="DP", ylab="CDF")
```

## Quality vs. coverage

```{r fig.width=8, fig.height=5}
vd <- data.frame(QUAL=qual(vcf),DP=info(vcf)$DP+dp.dither)
ggplot(vd, aes(DP,QUAL)) + geom_bin2d(bins=64) + scale_fill_gradient(trans="sqrt") + xlim(0,64) + ylim(0,2000)
```

## Overall metrics

Helper function to report metrics:

```{r}
vcf.stats <- function(the.vcf) {
  cat("N (biallelic) =", nrow(the.vcf), "\n")
  
  qq <- quantile(qual(the.vcf), probs=c(0.5,0.1,0.05,0.01))
  cat("QUAL: median =", qq[1], "pct10 =", qq[2], "pct5 =", qq[3], "pct1 =", qq[4], "\n")
  
  dpq <- quantile(info(the.vcf)$DP, probs=c(0.5,0.1,0.05,0.01))
  cat("DP: median =", dpq[1], "pct10 =", dpq[2], "pct5 =", dpq[3], "pct1 =", dpq[4], "\n")
  
  snvs <- as.character(as.list(info(the.vcf)$TYPE)) == "snp"
  snvs.het <- snvs & as.numeric(as.list(info(the.vcf)$AF)) == 0.5
  snvs.hom <- snvs & as.numeric(as.list(info(the.vcf)$AF)) == 1
  
  cat("SNVs:", sum(snvs), "=", sum(snvs.het), "het +", sum(snvs.hom), "hom\n")
  
  ref.nucs <- as.character(ref(the.vcf))[snvs]
  alt.nucs <- as.character(unlist(alt(the.vcf)))[snvs]
  transitions <- (sum(ref.nucs == 'A' & alt.nucs == 'G') + sum(ref.nucs == 'G' & alt.nucs == 'A')
                   + sum(ref.nucs == 'C' & alt.nucs == 'T') + sum(ref.nucs == 'T' & ref.nucs == 'C'))
  transversions <- (sum(ref.nucs == 'A' & alt.nucs == 'C') + sum(ref.nucs == 'C' & alt.nucs == 'A')
                   + sum(ref.nucs == 'A' & alt.nucs == 'T') + sum(ref.nucs == 'T' & alt.nucs == 'A')
                   + sum(ref.nucs == 'C' & alt.nucs == 'G') + sum(ref.nucs == 'G' & ref.nucs == 'C')
                   + sum(ref.nucs == 'G' & alt.nucs == 'T') + sum(ref.nucs == 'T' & ref.nucs == 'G'))
  
  cat("transitions =",transitions,"transversions =",transversions,"Ts/Tv =",transitions/transversions,"\n")
}
```

Compute the metrics over all biallelic sites:

```{r}
vcf <- vcf[elementLengths(info(vcf)$TYPE) == 1]
vcf.stats(vcf)
```

## Metrics by chromosome

```{r}
for (i in 1:length(chroms)) {
  cat("Statistics for",chroms[i],"\n")
  vcf.stats(keepSeqlevels(vcf,chroms[i]))
  cat("\n")
}
```
