Whole-genome VCF Quality Control
========================================================

Load variants on the canonical chromosomes (b37/hs37d5 chromosome naming assumed):

```{r message=FALSE}
library(VariantAnnotation)
library(ggplot2)
keep <- list()
keep$fixed <- c("ALT","QUAL")
keep$info <- c("TYPE","AF")
keep$geno <- c("GT","DP","RO","AO")
vcf <- readVcf("/input.vcf.gz","hs37d5",ScanVcfParam(fixed=keep$fixed,info=keep$info,geno=keep$geno))
vcf <- keepSeqlevels(vcf,c(as.character(1:22),"X","Y"))

cat("sites =", nrow(vcf), "alt. alleles =", sum(elementLengths(geno(vcf)$AO)))
```

## Quality distribution

```{r fig.width=6, fig.height=4, warning=FALSE}
qplot(qual(vcf), stat="ecdf", geom="step", xlim=c(0,2000), xlab="QUAL", ylab="CDF")
```

## Coverage distribution

```{r fig.width=6, fig.height=4, warning=FALSE}
qplot(as.vector(geno(vcf)$DP), stat="ecdf", geom="step", xlim=c(0,64), xlab="DP", ylab="CDF")
```

```{r fig.width=6, fig.height=4, warning=FALSE}
qplot(unlist(geno(vcf)$AO), stat="ecdf", geom="step", xlim=c(0,64), xlab="AO", ylab="CDF")
```

## Quality vs. coverage

```{r fig.width=8, fig.height=5}
my.expand <- function(vec,lengths) {
  unlist(lapply(1:length(lengths),function (i) { rep(vec[i],lengths[i]) } ))
}
vd <- data.frame(QUAL=my.expand(qual(vcf),elementLengths(geno(vcf)$AO)),AO=unlist(geno(vcf)$AO))

ggplot(vd, aes(AO,QUAL)) + geom_bin2d(bins=64) + scale_fill_gradient(trans="sqrt") + xlim(0,64) + ylim(0,2000)
```

## Overall metrics

Helper function to report metrics:

```{r}
vcf.stats <- function(chrom,the.vcf) {
  cat("Metrics for", chrom, "\n")
  cat("biallelic sites =", nrow(the.vcf), "\n")
  
  qq <- quantile(qual(the.vcf), probs=c(0.5,0.1,0.05,0.01))
  cat("QUAL: median =", qq[1], "pct10 =", qq[2], "pct5 =", qq[3], "pct1 =", qq[4], "\n")
  
  dpq <- quantile(as.vector(geno(the.vcf)$DP), probs=c(0.5,0.1,0.05,0.01))
  cat("DP: median =", dpq[1], "pct10 =", dpq[2], "pct5 =", dpq[3], "pct1 =", dpq[4], "\n")
  
  aoq <- quantile(unlist(geno(the.vcf)$AO), probs=c(0.5,0.1,0.05,0.01))
  cat("AO: median =", aoq[1], "pct10 =", aoq[2], "pct5 =", aoq[3], "pct1 =", aoq[4], "\n")
  
  snvs <- as.character(as.list(info(the.vcf)$TYPE)) == "snp"
  snvs.het <- snvs & as.numeric(as.list(info(the.vcf)$AF)) == 0.5
  snvs.hom <- snvs & as.numeric(as.list(info(the.vcf)$AF)) == 1
  
  cat("SNVs:", sum(snvs), "=", sum(snvs.het), "het +", sum(snvs.hom), "hom", "... % het =", 100*sum(snvs.het)/(sum(snvs.hom)+sum(snvs.het)), "\n")
  
  ref.nucs <- toupper(as.character(ref(the.vcf))[snvs])
  alt.nucs <- toupper(as.character(unlist(alt(the.vcf)))[snvs])
  transitions <- (sum(ref.nucs == 'A' & alt.nucs == 'G') + sum(ref.nucs == 'G' & alt.nucs == 'A')
                   + sum(ref.nucs == 'C' & alt.nucs == 'T') + sum(ref.nucs == 'T' & ref.nucs == 'C'))
  transversions <- (sum(ref.nucs == 'A' & alt.nucs == 'C') + sum(ref.nucs == 'C' & alt.nucs == 'A')
                   + sum(ref.nucs == 'A' & alt.nucs == 'T') + sum(ref.nucs == 'T' & alt.nucs == 'A')
                   + sum(ref.nucs == 'C' & alt.nucs == 'G') + sum(ref.nucs == 'G' & ref.nucs == 'C')
                   + sum(ref.nucs == 'G' & alt.nucs == 'T') + sum(ref.nucs == 'T' & ref.nucs == 'G'))
  
  cat("transitions =",transitions,"transversions =",transversions,"Ts/Tv =",transitions/transversions,"\n\n")
}
```

Compute the metrics over all biallelic sites:

```{r}
vcf <- vcf[elementLengths(info(vcf)$TYPE) == 1]
vcf.stats("all biallelic sites",vcf)
```

## Metrics for X, Y, and autosomes

```{r}
vcf.stats("X",keepSeqlevels(vcf,"X"))
vcf.stats("Y",keepSeqlevels(vcf,"Y"))
vcf <- keepSeqlevels(vcf,c(as.character(1:22)))
vcf.stats("autosomes",vcf)
```


## Metrics by autosome

```{r}
for (i in 1:22) {
  ci <- as.character(i)
  vcf.stats(ci,keepSeqlevels(vcf,ci))
}
```

